### PR TITLE
Discard merged any marker constraints

### PR DESCRIPTION
**Pull request by radoering**
Resolves an issue revealed in https://github.com/python-poetry/poetry/pull/6950#issuecomment-1327494436

Based on #4590, after adapting the marker of "any marker dependencies", these dependencies have to be discarded if the marker is empty, not compatible with the project's python constraint or (during install) not compatible with the environment.

- [x] Added **tests** for changed code.
- [ ] Updated **documentation** for changed code.

### CODE REVIEW

**Review by neersighted — CHANGES_REQUESTED** [review]

File: src/poetry/puzzle/provider.py
Comment (radoering) [reply src/poetry/puzzle/provider.py:1036]: `any_markers_dependencies` and `other_markers_dependencies` is used unconditionally in line 1000 and that's always required so we can't move the computation. Intersecting the constraints is always necessary, no matter if the inverted marker is empty or not.

Further, we can't use a set because dependencies can be equal even if they have different markers (and here we care about markers).

**Comment by dimbleby** [comment]
I have doubts about correctness, especially in cases where the requirements actually are incompatible.

For instance consider
```toml
requests = [
  { version = "=2.24.0" },
  { version = "=2.25.0", markers = "implementation_name == 'pypy'" }
]
```
the result of this is that 2.24.0 is chosen unconditionally, which one way or the other surely wasn't what the user intended.

It would certainly be simpler to make the user responsible for ensuring that their markers are non-overlapping, rather than trying to infer what they probably meant.

That is, perhaps the simpler approach is just to search for, and refuse to process, overlapping markers.

Alternatively poetry could search for overlapping markers and reject them if and only if they have contradictory constraints - but still do something like what this MR does when their constraints are compatible.  Feels like it might be getting a bit complicated though...

**Comment by radoering** [comment]
> I have doubts about correctness, especially in cases where the requirements actually are incompatible.

Good catch. (This has already been an issue before this PR.) I just added a fix for that, too.

> It would certainly be simpler to make the user responsible for ensuring that their markers are non-overlapping, rather than trying to infer what they probably meant.

It would, but I'd rather support compatible overlapping markers if there isn't any PEP that says overlapping markers are forbidden.

> also the special-casing of "any" bothers me. That's not the only way to write overlapping markers.

That's right. At the moment we are just dealing with one more common case of overlapping markers. Or more exactly with two cases because `_merge_dependencies_by_marker` is handling the special case of completely overlapping markers. I already have a follow up PR in mind to deal with overlapping markers in general, i.e. generalizing `_handle_any_marker_dependencies` and removing `_merge_dependencies_by_marker` (and probably also some of the code in between the two calls).

File: src/poetry/puzzle/provider.py
Comment (neersighted) [reply src/poetry/puzzle/provider.py:1055]: Okay, fair enough. I wonder if the equality situation here was a mistake (and we should include markers) -- but given we are deliberately rewriting markers, maybe not?

**Comment by neersighted** [comment]
In general, I think we're committed to this formulation of markers, as to do otherwise would be a breaking change, and likely break existing projects. I'm reasonably sure our extras implementation relies on part of this as well. Overall, I think trying to refactor this to be cleaner and better abstracted/architected over time is the only way forward.

File: src/poetry/puzzle/provider.py
Comment (radoering) [reply src/poetry/puzzle/provider.py:1055]: Difficult to say. Changing `Dependency.__eq__` to consider the `marker` property seems to break some downstream tests, but only because we don't care for markers in our expectations. Searching for usage of `Dependency` in sets or as key in dicts, I only found `Provider._deferred_cache` which might get a bit worse at caching if we consider markers. At a quick glance, I couldn't find any other obvious side effects but of course that doesn't mean it's safe. There can still be simple comparisons that rely on equality as is.

I don't think rewriting markers necessarily is a reason for not considering markers in equality. It's just a reason to not include it in the hash.

File: src/poetry/puzzle/provider.py
Comment (dimbleby) [inline src/poetry/puzzle/provider.py:1056]: doesn't this code have the same problem, then?

eg if the list is `[foo (with a marker that's not any), foo (with a marker that is any)]`, then this code will remove the first of those elements from the list, right?

perhaps there's some reason that can't happen?

**Comment by dimbleby** [comment]
this one makes myhead hurt! perhaps it would be sensible to unblock #6950 and therefore the 1.3.0 release by rewriting the markers in that one so that they don't need this fix?

File: src/poetry/puzzle/provider.py
Comment (radoering) [reply src/poetry/puzzle/provider.py]: Good catch. The workaround is non-sense. Actually, this case probably never occurs because we merge dependencies by constraint earlier. Nevertheless, I think it wouldn't be correct to use `remove` here. Now, it should be correct.

**Comment by radoering** [comment]
I was hoping that we were pretty far along with this review. Actually, this is just a small fix. Partial support for overlapping markers was not introduced here, but was already present before this PR. Anyway, IMO this doesn't really block the release as long as nobody has updated the changelog. 😉

**Comment by radoering** [comment]
That's right. I just mean reworking #6950 is just a matter of minutes but even with that the changelog has to be updated and that's the real work. We might as well wait with reworking #6950 until that's the only thing the release is waiting for. (As you have probably already noticed, I'd prefer to keep #6950 as is which requires this PR.)

**Comment by dimbleby** [comment]
> ... until that's the only thing the release is waiting for

So far as I can tell from https://github.com/python-poetry/poetry/issues/6972, #6950 _is_ the only thing the release is waiting for

File: src/poetry/puzzle/provider.py
Comment (dimbleby) [inline src/poetry/puzzle/provider.py:1059]: this is already `other_markers_dependencies` isn't it?

**Comment by dimbleby** [comment]
left a note suggesting a little tidying, otherwise looks good to me

File: src/poetry/puzzle/provider.py
Comment (radoering) [reply src/poetry/puzzle/provider.py]: It seems I was too focussed on what's not working to see the simplest solution.

**Review by neersighted — APPROVED** [review]
LGTM!

**Pull request merged**
